### PR TITLE
Fix MLABecLaplacian semi-coarsening line-solve smoother

### LINKED ISSUE #5692: MLABecLaplacian semicoarsening line-solve smoother: wrong x-hi mask in 2D, tile boxes under OpenMP, host kernel on GPU
**Severity:** high/medium · **Category:** gpu-parallel-mpi, memory-ub · **Subsystem:** LinearSolvers (blast radius 4/5)

**Locations:** `Src/LinearSolvers/MLMG/AMReX_MLABecLap_2D_K.H:374`, `Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:998`, `Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:1202`

**Based on commit `248fcfb7a5`** (branch `ai_audit`; line numbers refer to that tree).

All three sit on the `abec_gsrb_with_line_solve` path that `Fsmooth` takes for semicoarsened MG levels (`regular_coarsening == false`, no overset mask). The kernel was transcribed from 3D and never hardened for tiling or GPU, and no in-tree test enables semicoarsening on a cell-centred op, so none of it is exercised by CI.

## The defect

**`Src/LinearSolvers/MLMG/AMReX_MLABecLap_2D_K.H:374`** — 2D abec_gsrb_with_line_solve gates the x-hi boundary rho correction with the y-hi mask m3 instead of the x-hi mask m2 (cf2 at line 358 uses m2), reading m3(vhi.x+1,j,0) out of bounds of the y-hi mask strip on every box's last column.

**`Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:998`** — In GPU builds with semicoarsening (irregular coarsening) and no overset mask, Fsmooth's GPU gate excludes the case and control falls into the host block, which runs the host-only abec_gsrb_with_line_solve on Array4s pointing to device memory.

**`Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:1202`** — Fsmooth passes the tile box (tbx) from a tiling-enabled OpenMP MFIter to abec_gsrb_with_line_solve, which assumes box==validbox: tridiagonal lines are truncated at tile edges, segment-end reads race with same-color writes by threads owning adjacent tiles, and boundary masks are tested at valid-box edges for interior tile edges. _(also reported as F198, F199)_

## Why it matters

*F046:* AMREX_SPACEDIM=2, CPU: MLABecLaplacian with LPInfo().setSemicoarsening so amrlev==0/mglev>0 has irregular coarsening and dy<dx (dhy>dhx). Fsmooth calls the line solve; at i==vhi.x it reads m3(vhi.x+1,j,0), outside the mask FAB box [vlo.x..vhi.x]x{vhi.y+1} (extent_rad=0 since isCrossStencil). Debug builds abort via Array4 index_assert; release builds apply/skip the Dirichlet rho correction on garbage, corrupting the smoother and stalling MLMG.

*F204:* CUDA/HIP/SYCL build (device arena, amrex.the_arena_is_managed=0 default), MLABecLaplacian + LPInfo().setSemicoarsening(true), GS smoother: at amrlev 0 irregular mglev, Gpu::inLaunchRegion() is true but (overset||regular_coarsening) is false -> host MFIter loop dereferences device pointers in abec_gsrb_with_line_solve -> segfault (acknowledged only by the comment at line 1201, no guard).

*F205:* 3D CPU + OpenMP (threads>1), default mfiter_tile_size (1024000,8,8) chops the z/y solve direction into 8-cell segments; MLABecLaplacian with LPInfo().setSemicoarsening(true), GS smoother: thread A reads phi(i,j,lo.z-1) (AMReX_MLABecLap_3D_K.H:559) while thread B writes that same-color cell -> data race; when the box touches a physical z boundary, interior tile segments drop the neighbor coupling entirely -> nondeterministic results and MLMG convergence failure.

## How to reach it

- *F046:* config: AMReX_SPACEDIM=2, CPU (any MPI/OMP; line solve is CPU-only). gcc.yml/clang.yml build 2D incl. Debug, but no ctest input enables semicoarsening.
- *F046:* API path: MLABecLaplacian op({geom},{ba},{dm}, LPInfo().setSemicoarsening(true).setMaxSemicoarseningLevel(4).setSemicoarseningDirection(1)); op.setDomainBC(...); op.setScalars(a,b); op.setACoeffs/setBCoeffs; MLMG(op).solve(...). At amrlev 0, mglev>0 mg_coarsen_ratio_vec=(2,1) => regular_coarsening=false, dhy>dhx => abec_gsrb_with_line_solve; every box's i==vhi.x column reads m3(vhi.x+1,j,0). Also reached by the default heuristic when y stops coarsening first (e.g. 128x8 domain).
- *F046:* test coverage: Tests/LinearSolvers/ABecLaplacian_C (ParmParse `semicoarsening=1 max_semicoarsening_level=N`); ctest runs only inputs-rt-poisson-lev (no semicoarsening) and asserts nothing on the solution.
- *F046:* downstream use: none found for the affected path. WarpX enables semicoarsening only on nodal ops: ablastr/fields/PoissonSolver.H:312-314 (MLEBNodeFDLaplacian :323), VectorPoissonSolver.H:164-166 (MLEBNodeFDLaplacian :170), Initialization/DivCleaner/ProjectionDivCleaner.cpp:205 (MLNodeLaplacian). incflo/AMReX-Hydro/ERF use MLABecLaplacian with default LPInfo() (e.g. AMReX-Hydro/Projections/hydro_MacProjector.H:82). REMORA: none.
- *F204:* config: AMReX_GPU_BACKEND=CUDA|HIP|SYCL, AMReX_SPACEDIM=2 or 3, default amrex.the_arena_is_managed=0; cuda.yml/hip.yml/intel.yml build such configs but no GPU job runs ctest.
- *F204:* API path: LPInfo info; info.setSemicoarsening(true).setMaxSemicoarseningLevel(4); MLABecLaplacian op({geom},{ba},{dm},info); op.setScalars(1,1); op.setACoeffs(0,a); op.setBCoeffs(0,b); MLMG mlmg(op); mlmg.solve(...) on e.g. a 64x64x8 domain so one direction stops coarsening. Public API.
- *F204:* test coverage: Tests/LinearSolvers/ABecLaplacian_C reaches it only with ParmParse semicoarsening=1 max_semicoarsening_level=N (MyTest.cpp:161-163); no shipped inputs enable it and it asserts nothing.
- *F205:* config: AMReX_SPACEDIM=3, CPU build (no GPU backend), AMReX_OMP=ON for the race (mask/truncation defect also serial). gcc.yml builds two -DAMReX_OMP=ON jobs that run ctest; none enable semicoarsening.
- *F205:* API path: LPInfo info; info.setSemicoarsening(true).setMaxSemicoarseningLevel(8).setSemicoarseningDirection(2); MLABecLaplacian lp(geom, grids, dmap, info); // 3D, box z-length in (8,32] lp.setDomainBC(lobc, hibc); // non-periodic in z lp.setScalars(a,b); lp.setACoeffs(0,acoef); lp.setBCoeffs(0,bcoef); MLMG mlmg(lp); mlmg.solve({&sol},{&rhs},1e-10,0); // OMP_NUM_THREADS>1
- *F205:* test coverage: Tests/LinearSolvers/ABecLaplacian_C with `semicoarsening=1 max_semicoarsening_level=N` reaches Fsmooth line solve; prints residual norms only, no assert; no in-tree test sets semicoarsening_direction.

## Suggested fix

Three independent changes; land the 2D mask fix first (pure correctness, no behaviour question), then the two `Fsmooth` dispatch changes. **Kernel (F046).** In `AMReX_MLABecLap_2D_K.H` the x-hi rho correction must test the x-hi mask `m2`, exactly as `cf2` does four lines earlier. The `m3` is a leftover from the 3D twin (`AMReX_MLABecLap_3D_K.H:619`), where the six-mask Orientation numbering makes `m3` x-hi; in 2D `Fsmooth` passes `m3 = maskvals[3]` = y-hi, whose FAB box is the strip `{vhi.y+1}`, so `m3(vhi.x+1,j,0)` is out of bounds. **Tiling (F205).** The kernel assumes `box == vbox`: the tridiagonal line spans `lo..hi` of `box`, while the boundary-mask tests and the rho corrections use `vlo/vhi`. With `mfi_info.EnableTiling()` and the default CPU tile (8 cells in y/z) that assumption is false, so lines are chopped into segments, interior segment ends read same-colour cells owned by another thread's tile, and the mask test at the valid-box edge silently drops the neighbour coupling. The right fix is to not tile when this branch will be taken: build `mfi_info` without `EnableTiling()` when `!regular_coarsening && !m_overset_mask[amrlev][mglev]`, so `tbx == vbx` and each OpenMP thread owns whole boxes. Other branches (overset `abec_gsrb_os`, regular coarsening, GPU) keep tiling. The GS/Jacobi kernels are pointwise and unaffected. Maintainer decision: once lines span the full valid box, the kernel's hard-wired `ilen > 32` Abort fires on any box longer than 32 cells in the solve direction, whereas today tiling hid it (and produced wrong answers instead). Either accept that limit (it already exists, and `max_grid_size` defaults to 32 in 2D/3D... the semicoarsened direction is the fine one, so boxes can exceed it) or raise the `Array1D<T,0,31>` bound. Also confirm that overset + semicoarsening is intentionally routed to `abec_gsrb_os` rather than the line solve. **GPU (F204).** The GPU gate `Gpu::inLaunchRegion() && (overset || regular_coarsening)` deliberately excludes the line-solve case, but the fall-through host loop then runs the host-only kernel on device pointers. Until a device line solve exists, replace the silent segfault with an `AMREX_ALWAYS_ASSERT_WITH_MESSAGE` inside the existing `#ifdef AMREX_USE_GPU` block, mirroring the gate's condition, so users get a clear "line solve not supported on GPU" message. CPU builds are untouched. Writing a device kernel (one thread per transverse line, like the nodal semicoarsening path) would be the full fix and is left to the maintainer. Suggest adding a `semicoarsening=1` + `semicoarsening_direction` input to `Tests/LinearSolvers/ABecLaplacian_C` and registering it with ctest so this path is built in Debug and run under OpenMP.

For `Src/LinearSolvers/MLMG/AMReX_MLABecLap_2D_K.H:374` (F046):

```diff
--- a/Src/LinearSolvers/MLMG/AMReX_MLABecLap_2D_K.H
+++ b/Src/LinearSolvers/MLMG/AMReX_MLABecLap_2D_K.H
@@ -371,6 +371,6 @@ void abec_gsrb_with_line_solve (
                     if (i == vlo.x && m0(vlo.x-1,j,0) > 0) {
                         rho -= dhx*bX(i  ,j,0,n)*phi(i-1,j,0,n);
                     }
-                    if (i == vhi.x && m3(vhi.x+1,j,0) > 0) {
+                    if (i == vhi.x && m2(vhi.x+1,j,0) > 0) {
                         rho -= dhx*bX(i+1,j,0,n)*phi(i+1,j,0,n);
                     }
```

<sub>Gate the x-hi rho correction with the x-hi mask m2, matching cf2 on line 358 and the caller (MLABecLaplacian.H:1205 passes m2 = maskvals[2] as the 2D x-hi mask). The m3 came from the 3D idir==1 sibling (AMReX_MLABecLap_3D_K.H:619), where m3 is x-hi under the six-mask numbering; in 2D m3 is y-hi and m3(vhi.x+1,j,0) is out of its FAB box.</sub>

For `Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:998` (F204):

```diff
--- a/Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H
+++ b/Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H
@@ -995,6 +995,9 @@
     const RT alpha = m_a_scalar;
 
 #ifdef AMREX_USE_GPU
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(regular_coarsening || this->m_overset_mask[amrlev][mglev]
+                                     || Gpu::notInLaunchRegion(),
+                                     "MLABecLaplacian::Fsmooth: line solve is not supported on GPU");
     if (Gpu::inLaunchRegion()
         && (this->m_overset_mask[amrlev][mglev] || regular_coarsening))
     {
```

<sub>Turns the silent fall-through into the host line-solve on device memory into a clear abort, kept inside the existing #ifdef AMREX_USE_GPU block so CPU builds are untouched. Condition mirrors the GPU gate on the following if. A device line solve would be the full fix; until then semicoarsening without overset masks is rejected on GPU rather than segfaulting.</sub>

For `Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H:1202` (F205):

```diff
--- a/Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H
+++ b/Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H
@@ -1101,7 +1101,11 @@
 #endif
     {
         MFItInfo mfi_info;
-        mfi_info.EnableTiling().SetDynamic(true);
+        // The line solve used for semicoarsening needs whole valid boxes,
+        // not tiles.
+        if (regular_coarsening || this->m_overset_mask[amrlev][mglev]) {
+            mfi_info.EnableTiling().SetDynamic(true);
+        }
 
 #ifdef AMREX_USE_OMP
 #pragma omp parallel if (Gpu::notInLaunchRegion())
```

<sub>Disables MFIter tiling exactly when the CPU path will take the abec_gsrb_with_line_solve branch (no overset mask, irregular/semi-coarsening), so tbx == vbx and each tridiagonal line spans the full valid box; OpenMP threads then own whole boxes, removing the cross-tile same-color race and the mask-at-valid-edge mismatch. Other branches (overset, regular coarsening, GPU) keep tiling unchanged. Maintainer may verify the overset+semicoarsening combination is intentionally routed to abec_gsrb_os.</sub>

<sub>Diff(s) are against `248fcfb7a5`, written from the current source and verified only with `git apply --check` — never compiled, never run, never applied to the tree. Treat them as precise intent, not tested patches.</sub>

## Verification evidence

<details><summary><code>F046</code> — confirmed (two independent verifier lenses)</summary>

Lens 1 (refutation attempt): AMReX_MLABecLap_2D_K.H:374 `if (i == vhi.x && m3(vhi.x+1,j,0) > 0)` while cf2 (line 358) and all siblings use m2 (x-hi). Fsmooth passes m2=maskvals[2] (x-hi), m3=maskvals[3] (y-hi). Mask FAB box from BATbndryReg(face,cell,in_rad=0,out_rad=1,extent_rad=0; isCrossStencil()=true, MLCellLinOp.H:694) is [vlo.x..vhi.x]x{vhi.y+1}; offset = nx*(j-vhi.y) is negative for every j<vhi.y -> read before the allocation; AMREX_DEBUG index_assert aborts. Present since a19fb05e6c (#1072); no fix on development, not in KNOWN_STATE.

Lens 2 (reachability/intent): L374 `m3(vhi.x+1,j,0)` vs L358 cf2 `m2(vhi.x+1,j,0)`; Fsmooth (AMReX_MLABecLaplacian.H:1202) passes m2=x-hi, m3=y-hi. In 3D m3 IS x-hi (Orientation order), so this is a 3D transcription slip, verbatim since a19fb05e6c (#1072; PR body only says 'allow cell-center solver to do semi-coarsening'). `git show development:...2D_K.H` line 374 unchanged. Own public-API repro: Debug aborts '(16,0,0,0) is out of bound (0:15,16:16,0:0,0:0)'; Release+ASan heap-buffer-overflow READ size 4 at :374 via Fsmooth:1202/smooth:1552/mgVcycle:1770; plain Release happened to converge. Recompiling AMReX_MLMG.cpp with m2 converges (13 iters, identical solution).

</details>

<details><summary><code>F204</code> — confirmed (one verifier lens)</summary>

Lens 1 (refutation attempt): Fsmooth 998-999: `if (Gpu::inLaunchRegion() && (this->m_overset_mask[amrlev][mglev] || regular_coarsening))`; otherwise the host MFIter block runs and its final else (1199-1210) executes `// line solve does not with with GPU abec_gsrb_with_line_solve(tbx, solnfab, rhsfab, ...)` — a plain host function (MLABecLap_2D_K.H:308, 3D_K.H:452) dereferencing Array4s from sol.array(mfi) in The_Arena device memory (AMReX_Arena.cpp:59 the_arena_is_managed=false). regular_coarsening is false only for semicoarsened levels (mg_coarsen_ratio_vec, MLLinOp.H:1253-1280). Original a19fb05e6c had `Gpu::LaunchSafeGuard lsg(false); // xxxxx gpu todo` — unfinished TODO, no Abort/guard added since. Not fixed on development.

</details>

<details><summary><code>F205</code> — confirmed (one verifier lens); duplicates folded in: F198, F199</summary>

Lens 1 (refutation attempt): MLABecLaplacian.H:1104 `mfi_info.EnableTiling().SetDynamic(true)`; :1202 passes `tbx` as the kernel `box` and `vbx` as `vbox`. 3D_K.H:559 `if (k == lo.z) { a_ls=0; if (!(m2(i,j,vlo.z-1) > 0)) r_ls += dhz*bZ*phi(i,j,k-1,n); }` - at an interior tile edge (lo.z != vlo.z) it reads the same-(i,j)-color cell owned by the adjacent tile/thread and tests the mask at the valid-box edge, dropping the coupling when the box touches a non-periodic boundary. FabArrayBase.cpp:49 default CPU 3D tile (1024000,8,8). Tile box has been passed since PR #1072; no dev fix (`git log ai_audit..development`). Issue #5000 covers only the Abort checks.

</details>

---

<sub>Based on commit `248fcfb7a5`, the tree the audit verified against. From an automated audit of AMReX `Src/`. Audit finding ids: F046, F204, F205, F198, F199. Reviewer unit(s): LS/ABecLap-1, LS/ABecLap-2, sweep:dim-mirrors-A. Nothing here was compiled or run except where the evidence says so — the failure scenarios are code reasoning, so the reaching configuration above is the cheapest way to confirm or refute it.</sub>

### PR DESCRIPTION
**Pull request by WeiqunZhang**
The 2D kernel gated the x-hi rho correction with the y-hi mask, reading out of bounds on every box's last column.  Fsmooth also passed tile boxes to a kernel that assumes box == validbox, truncating the tridiagonal lines and racing between OpenMP threads, and fell through to the host-only kernel on GPU.  Tiling is now disabled for this branch, the 32-cell line limit is gone, and GPU builds abort with a clear message.  New ctest covers it.

Fixes #5692.

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
# PR #5817: Fix MLABecLaplacian semi-coarsening line-solve smoother

PR: https://github.com/AMReX-Codes/amrex/pull/5817
Reviewed commit: `1c823b835d9f683b9646015e11298b5b4e26ce3e`
Reviewed at: 2026-09-07 20:32 UTC
Status: Complete
Result: No actionable findings

## Findings

No actionable findings. The 2D high-X mask uses the correct face; the line branch runs on whole boxes; dynamic scratch removes the 32-cell limit without introducing SIMD sharing. The new test has a separate working directory.

## Validation

Inspected PR description, conversation, inline comments, and the complete diff with `git diff ae079342654b60f95effddd4377cd9ffaf573f5f 1c823b835d9f683b9646015e11298b5b4e26ce3e`. Built an exact-head archive in /tmp/amrex-pr-review/5817/src with `cmake -S /tmp/amrex-pr-review/5817/src -B /tmp/amrex-pr-review/5817/build -DAMReX_SPACEDIM=2 -DAMReX_MPI=ON -DAMReX_OMP=OFF -DAMReX_FORTRAN=OFF -DAMReX_PARTICLES=OFF -DAMReX_AMRLEVEL=OFF -DCMAKE_BUILD_TYPE=Debug -DAMReX_ENABLE_TESTS=ON -DAMReX_TEST_TYPE=Small`, then `cmake --build /tmp/amrex-pr-review/5817/build -j8 --target Test_LinearSolvers_ABecLaplacian_C_semicoarsening_2d`. Build passed. `ctest --test-dir /tmp/amrex-pr-review/5817/build -R '^LinearSolvers_ABecLaplacian_C_semicoarsening_2d$' --output-on-failure` passed on two MPI ranks. From the test working directory, `timeout 30 mpiexec -n 2 ./Test_LinearSolvers_ABecLaplacian_C_semicoarsening_2d inputs-rt-abeclap-semicoarsening max_grid_size=64` also passed (17 iterations; relative residual 6.51e-11). Inspected all dimensional changes and generated CTest registration. No 3D, OpenMP, or GPU run was performed.

CI at reviewed head: SUCCESS: 75. CI status is separate from the source assessment.

Reviewed by gpt-6-astra.

**Review by asalmgren — APPROVED** [review]
